torch/autograd/functional.py

```diff
@@ -0,0 +1,714 @@
+import torch
+
+# Utility functions
+
+def _as_tuple(inp, arg_name, fn_name):
+    # Ensures that the inputs is a tuple of Tensors
+    # Returns whether or not the original inp was a tuple and the tupled version of the input
+    tuple_inp = True
+    if not isinstance(inp, tuple):
+        inp = (inp,)
+        tuple_inp = False
+
+    for i, el in enumerate(inp):
+        if not torch.is_tensor(el):
+            raise TypeError("The {} given to {} must be either a Tensor or a tuple of Tensors but the"
+                            " value at index {} has type {}.".format(arg_name, fn_name, i, type(el)))
+
+    return tuple_inp, inp
+
+def _tuple_postprocess(inp, tuple_inp):
+    # Unpacks a potentially nested tuple of Tensors
+    # It is used to invert _as_tuple before returning to the user
+    if isinstance(tuple_inp, tuple):
+        assert len(tuple_inp) == 2
+        if not tuple_inp[1]:
+            inp = tuple(el[0] for el in inp)
+        if not tuple_inp[0]:
+            inp = inp[0]
+    else:
+        if not tuple_inp:
+            inp = inp[0]
+    return inp
+
+def _grad_preprocess(inputs, create_graph, need_graph):
+    # Preprocess the inputs to make sure they require gradient
+    # inputs is a tuple of Tensor to preprocess
+    # create_graph specifies if the user wants gradients to flow back to the original input
+    # need_graph specifies if we internally want gradients to flow back for this Tensor
```

Comment:
Which Tensor? `inputs`?

Reply:
For the corresponding entry that is in `res`. I'll make that clearer in the comment.